presto-main/src/main/java/com/facebook/presto/sql/analyzer/StatementAnalyzer.java

```diff
@@ -1220,6 +1234,17 @@ protected Scope visitSetOperation(SetOperation node, Optional<Scope> scope)
             return createAndAssignScope(node, scope, outputDescriptorFields);
         }
 
+        private boolean isExpensiveDistinctUnion(SetOperation setOperation, Type[] outputTypes)
+        {
+            return setOperation instanceof Union && setOperation.isDistinct() && outputTypes.length > UNION_DISTINCT_FIELDS_WARNING_THRESHOLD
+                    && Arrays.stream(outputTypes).anyMatch(this::isComplexType);
```

Comment:
maybe just do `Arrays.stream(outoutTypes).anyMatch(type -> type instanceof RealType || type instanceof DoubleType || type instanceof MapType || type instanceof ArrayType)` instead. Normally we don't put simple logic in separate functions if they are not reused much. Also REAL and DOUBLE are not really complex types. Oh, maybe also add `RowType` to the check.